cubejsclient/filters.py

```diff
@@ -0,0 +1,170 @@
+
+        Returns:
+            list of (start date, end date) or the relative date range
+
+        """
+        if self.start_date is not None and self.end_date is not None:
+            return [self.start_date, self.end_date]
+        elif self.relative:
+            return self.relative
+        else:
+            raise ValueError("Expected (start date, end date) or relative")
+
+
+class TimeDimension:
+    """Represents the combo of a time dimension and filter"""
+
+    def __init__(
+        self,
+        dimension: Dimension,
+        date_range: DateRange,
+        compare_date_range: Optional[List[DateRange]] = None,
+        granularity: TimeGranularity = TimeGranularity.null,
+    ) -> None:
+        """Initializer
+
+        Args:
+            dimension: Time dimension reference
+            date_range: Absolute or relative date range
+            compare_date_range: An array of date ranges to compare a measure change over
+                previous period
+            granularity: A granularity for a time dimension
+
+        """
+        self.dimension = dimension
+        self.date_range = date_range
+        self.compare_date_range = compare_date_range
+        self.granularity = granularity
+
+    def serialize(self) -> Dict[str, Any]:
+        """Serialize the object for building a query
+
+        Returns:
+            dict of properties
+
+        """
+        result = {
+            "dimension": self.dimension.serialize(),
+            "dateRange": self.date_range.serialize(),
+            "granularity": self.granularity.value,
+        }
+        if self.compare_date_range:
+            result["compareDateRange"] = [
+                dr.serialize() for dr in self.compare_date_range
+            ]
+
+        return result
+
+
+class Filter:
+    """Represents a data filter"""
+
+    def __init__(
+        self,
+        member: Union[Dimension, Measure],
+        operator: FilterOperator,
+        values: Optional[List[Any]] = None,
+    ) -> None:
+        """Initializer
+
+        Args:
+            member: Dimension or measure to be used in the filter
+                * When you filter on a dimension, you are restricting the raw data
+                  before any calculations are made.
+                * When you filter on a measure, you are restricting the results after
+                  the measure has been calculated.
+            operator: An operator to be used in the filter.
+            values: An array of values for the filter.
+
+        """
+        self.member = member
+        self.operator = operator
+        self.values = values
+
+    def serialize(self) -> Dict[str, Any]:
+        """Serialize the object for building a query
+
+        Returns:
+            dict of properties
+
+        """
+        return {
+            "member": self.member.serialize(),
+            "operator": self.operator.value,
+            "values": [str(value) for value in self.values],
+        }
+
+
+class BooleanOperator:
```

Comment:
I think this might better be called `BooleanExpression` if it represents both the operator and the operands. Are you explicitly trying to do this differently from Lexio, where we have a BooleanExpression that is initialized with both an operator and the operands (in that case, conditions I think, although I'm not sure I like that design)?

Reply:
Expression makes sense. Not trying to do anything different than Lexio, just wanting to match the Cube.js API spec